### PR TITLE
Allow local port control on net_socket connections

### PR DESCRIPTION
**Pull request by gudinoff**
Allow to set and retrieve the local port of socket based network connections.
Project used to validate developments: [validation-test.zip](https://github.com/godotengine/godot/files/5455974/validation-test.zip)

### CODE REVIEW

File: core/io/stream_peer_tcp.cpp [hunk 1]
```diff
@@ -65,12 +65,15 @@ void StreamPeerTCP::accept_socket(Ref<NetSocket> p_sock, IP_Address p_host, uint
 
 	peer_host = p_host;
 	peer_port = p_port;
+	_sock->get_address(local_port);
 }
 
-Error StreamPeerTCP::connect_to_host(const IP_Address &p_host, uint16_t p_port) {
+Error StreamPeerTCP::connect_to_host(const IP_Address &p_host, int p_port, int p_local_port) {
```
Comment (jonbonazza) [inline core/io/stream_peer_tcp.cpp:71]: Why did you change the port data type from uint16_t to int? The whole port range fits in a short integer, so we should use it.

File: core/io/tcp_server.h [hunk 2]
```diff
@@ -44,11 +44,16 @@ class TCP_Server : public Reference {
...
+	uint16_t local_port = 0;
+	IP_Address local_address;
+
 	Ref<NetSocket> _sock;
 	static void _bind_methods();
 
 public:
 	Error listen(uint16_t p_port, const IP_Address &p_bind_address = IP_Address("*"));
+	Error get_address(IP_Address &local_address, uint16_t &local_port) const;
+	int get_local_port() const;
```
Comment (jonbonazza) [inline core/io/tcp_server.h:56]: Should return uint16_t instead of int

File: drivers/unix/net_socket_posix.cpp [hunk 3]
```diff
@@ -187,24 +187,31 @@ NetSocketPosix::NetError NetSocketPosix::_get_socket_error() const {
...
+	if (err == WSAEADDRINUSE || err == WSAEADDRNOTAVAIL)
+		return ERR_NET_ADDRESS_INVALID_OR_UNAVAILABLE;
+	if (err == WSAEACCES)
+		return ERR_NET_UNAUTHORIZED;
 	print_verbose("Socket error: " + itos(err));
 	return ERR_NET_OTHER;
 #else
-	if (errno == EISCONN) {
+	// More info here https://man7.org/linux/man-pages/man2/bind.2.html
+	if (errno == EISCONN)
```
Comment (jonbonazza) [inline drivers/unix/net_socket_posix.cpp:205]: I don't know that this is allowed by the Godot style guidelines. @NathanLovato opinions?

Either way, Im not a fan of changing existing code style willy nilly. It convolutes the PR unnecessarily.

Comment (akien-mga) [reply drivers/unix/net_socket_posix.cpp:205]: That's a wrong change indeed, braces are required.

File: core/io/stream_peer_tcp.cpp (hunk 1 above)
Comment (dam-coder) [reply core/io/stream_peer_tcp.cpp:71]: My choice to use `int` instead of `uint16_t` was to mimic what I saw in other functions, for example:
- `Error PacketPeerUDP::connect_to_host(const IP_Address &p_host, int p_port)`;
- `Error PacketPeerUDP::listen(int p_port, const IP_Address &p_bind_address, int p_recv_buffer_size)`.

I assumed all methods that were _binded_ (accessible from the script side) would use `int` instead of `uint16_t`.

File: core/io/tcp_server.h (hunk 2 above)
Comment (dam-coder) [reply core/io/tcp_server.h:56]: Please see my [previous answer](https://github.com/godotengine/godot/pull/43174/files/4548b790985cdd0adbb6d5fb2f7498584ecc2f55#r528388372) to a similar comment.
Again, I see no problem in changing this to `uint16_t`, I just want to double check.

File: drivers/unix/net_socket_posix.cpp (hunk 3 above)
Comment (dam-coder) [reply drivers/unix/net_socket_posix.cpp:205]: My personal preference is to always use braces on `if` blocks.
In this case, I was just trying to follow the same code style that was used in the lines above.
I'll revert this change.

**Comment by dam-coder** [comment]
Is there anything else to address?

**Comment by Faless** [comment]
What is the purpose of `get_local_port` in UDP/TCP server? They are set explicitly in listen...

File: core/io/net_socket.h [hunk 4]
```diff
@@ -67,6 +67,9 @@ class NetSocket : public Reference {
 
 	virtual bool is_open() const = 0;
 	virtual int get_available_bytes() const = 0;
+	virtual Error get_address(IP_Address &local_address, uint16_t &local_port) const = 0;
+	virtual Error get_address(IP_Address &local_address) const = 0;
+	virtual Error get_address(uint16_t &local_port) const = 0;
```
Comment (Faless) [inline core/io/net_socket.h:72]: probably worth joining in a single method

File: drivers/unix/net_socket_posix.cpp [hunk 5]
```diff
@@ -716,6 +728,53 @@ int NetSocketPosix::get_available_bytes() const {
...
+	ERR_FAIL_COND_V(!is_open(), FAILED);
+
+	struct sockaddr_storage saddr;
+	socklen_t len = sizeof(saddr);
+	if (getsockname(_sock, (struct sockaddr *)&saddr, &len) == -1) {
+		print_verbose("Error when checking local socket address.");
+		return FAILED;
+	}
+
+	switch (saddr.ss_family) {
```
Comment (Faless) [inline drivers/unix/net_socket_posix.cpp:753]: should use `_set_ip_port`

**Review by Faless — CHANGES_REQUESTED** [review]
see comments, overall, this seems much more code then needed, the ENet bit should be moved to a separate PR

File: core/io/net_socket.h (hunk 4 above)
Comment (Faless) [reply core/io/net_socket.h:72]: And renamed to `getsockname` for consistency

File: drivers/unix/net_socket_posix.cpp (hunk 5 above)
Comment (dam-coder) [reply drivers/unix/net_socket_posix.cpp:753]: Makes sense. Will change it.

File: core/io/net_socket.h (hunk 4 above)
Comment (dam-coder) [reply core/io/net_socket.h:72]: In my understanding, joining in a single method will require to change the parameters to pointers (so they can be optional), and then call the method with NULLs on the parameters we are not interested in.
Is this the intended result of your suggestion, or is there another approach?

Renaming to `getsockname` seems to go against all the other methods naming convention and it will overlap with the method from `sys/socket.h` (I personally try to avoid this). May I suggest to rename it to `get_socket_address` or `get_socket_name`?

**Comment by dam-coder** [comment]
> What is the purpose of `get_local_port` in UDP/TCP server? They are set explicitly in listen...

I thought that, from the end coder perspective, it could be nice to simply get that info directly from the server instance instead of having to keep it on a separate variable.
To be honest... I wasn't totally convinced if I should include these or not, so I just left them and awaited for the reviewer's opinion.
What say you? :)

File: core/io/net_socket.h (hunk 4 above)
Comment (dam-coder) [reply core/io/net_socket.h]: Joined into a single method and renamed to `get_socket_address`.

**Comment by dam-coder** [comment]
In order to meet the requests made by @Faless  I've changed the following:
- Merge `get_address` methods into single method.
- Rename method `get_address` to `get_socket_address`.
- Use `_set_ip_port` instead of repeating code.
- Simplified implementation by removing cached values of `local_ip` and `local_port`.
- Moved ENet changes to a separate dependent PR -> #47220

I also agree with the comment "_this seems much more code then needed_", but the only way I found to reduce the code was to make the `Ref<NetSocket>` public so we don't need to create those "pipe methods" that only call the inner `_sock->get_socket_address(...)`. I assumed this was not the desired approach so I didn't go that way. Please let me know of any suggestions on how to achieve this. =)

File: core/io/packet_peer_udp.cpp [hunk 6]
```diff
@@ -186,6 +187,7 @@ Error PacketPeerUDP::listen(int p_port, const IP_Address &p_bind_address, int p_
 		return err;
 	}
 	rb.resize(nearest_shift(p_recv_buffer_size));
+
```
Comment (Faless) [inline core/io/packet_peer_udp.cpp:190]: remove extra line

File: core/io/packet_peer_udp.cpp [hunk 7]
```diff
@@ -206,10 +208,12 @@ void PacketPeerUDP::disconnect_shared_socket() {
 	close();
 }
 
-Error PacketPeerUDP::connect_to_host(const IP_Address &p_host, int p_port) {
+Error PacketPeerUDP::connect_to_host(const IP_Address &p_host, int p_port, int p_local_port) {
 	ERR_FAIL_COND_V(udp_server, ERR_LOCKED);
 	ERR_FAIL_COND_V(!_sock.is_valid(), ERR_UNAVAILABLE);
 	ERR_FAIL_COND_V(!p_host.is_valid(), ERR_INVALID_PARAMETER);
+	ERR_FAIL_COND_V_MSG(p_port < 0 || p_port > 65535, ERR_INVALID_PARAMETER, "The server port number must be set between 0 and 65535 (inclusive).");
+	ERR_FAIL_COND_V_MSG(p_local_port < 0 || p_local_port > 65535, ERR_INVALID_PARAMETER, "The local port number must be set between 0 and 65535 (inclusive).");
```
Comment (Faless) [inline core/io/packet_peer_udp.cpp:216]: remove `set`:
`The local port number must be between 0 and 65535 (inclusive).`

File: core/io/stream_peer_tcp.cpp [hunk 8]
```diff
@@ -67,10 +67,12 @@ void StreamPeerTCP::accept_socket(Ref<NetSocket> p_sock, IP_Address p_host, uint
 	peer_port = p_port;
 }
 
-Error StreamPeerTCP::connect_to_host(const IP_Address &p_host, uint16_t p_port) {
+Error StreamPeerTCP::connect_to_host(const IP_Address &p_host, int p_port, int p_local_port) {
 	ERR_FAIL_COND_V(!_sock.is_valid(), ERR_UNAVAILABLE);
 	ERR_FAIL_COND_V(_sock->is_open(), ERR_ALREADY_IN_USE);
 	ERR_FAIL_COND_V(!p_host.is_valid(), ERR_INVALID_PARAMETER);
+	ERR_FAIL_COND_V_MSG(p_port < 0 || p_port > 65535, ERR_INVALID_PARAMETER, "The server port number must be set between 0 and 65535 (inclusive).");
+	ERR_FAIL_COND_V_MSG(p_local_port < 0 || p_local_port > 65535, ERR_INVALID_PARAMETER, "The local port number must be set between 0 and 65535 (inclusive).");
```
Comment (Faless) [inline core/io/stream_peer_tcp.cpp:75]: same as above, remove `set`.

File: core/io/packet_peer_udp.cpp [hunk 9]
```diff
@@ -328,21 +338,32 @@ int PacketPeerUDP::get_packet_port() const {
 	return packet_port;
 }
 
+int PacketPeerUDP::get_local_port() const {
+	uint16_t local_port;
+	_sock->get_socket_address(NULL, &local_port);
```
Comment (Faless) [inline core/io/packet_peer_udp.cpp:343]: `NULL -> nullptr`

File: core/io/stream_peer_tcp.cpp [hunk 10]
```diff
@@ -300,11 +308,17 @@ IP_Address StreamPeerTCP::get_connected_host() const {
...
 
-uint16_t StreamPeerTCP::get_connected_port() const {
+int StreamPeerTCP::get_connected_port() const {
 	return peer_port;
 }
 
-Error StreamPeerTCP::_connect(const String &p_address, int p_port) {
+int StreamPeerTCP::get_local_port() const {
+	uint16_t local_port;
+	_sock->get_socket_address(NULL, &local_port);
```
Comment (Faless) [inline core/io/stream_peer_tcp.cpp:317]: `NULL -> nullptr`

File: core/io/tcp_server.cpp [hunk 11]
```diff
@@ -74,6 +76,16 @@ Error TCP_Server::listen(uint16_t p_port, const IP_Address &p_bind_address) {
 	return OK;
 }
 
+Error TCP_Server::get_socket_address(IP_Address *r_ip, uint16_t *r_port) const {
+	return _sock->get_socket_address(r_ip, r_port);
+}
+
+int TCP_Server::get_local_port() const {
+	uint16_t local_port;
+	_sock->get_socket_address(NULL, &local_port);
```
Comment (Faless) [inline core/io/tcp_server.cpp:85]: `NULL -> nullptr`

File: core/io/udp_server.cpp [hunk 12]
```diff
@@ -112,11 +114,19 @@ Error UDPServer::listen(uint16_t p_port, const IP_Address &p_bind_address) {
...
 	return OK;
 }
 
+Error UDPServer::get_socket_address(IP_Address *r_ip, uint16_t *r_port) const {
+	return _sock->get_socket_address(r_ip, r_port);
+}
+
+int UDPServer::get_local_port() const {
+	uint16_t local_port;
+	_sock->get_socket_address(NULL, &local_port);
```
Comment (Faless) [inline core/io/udp_server.cpp:126]: `NULL -> nullptr`

File: drivers/unix/net_socket_posix.cpp [hunk 13]
```diff
@@ -716,6 +728,28 @@ int NetSocketPosix::get_available_bytes() const {
 	return len;
 }
 
+Error NetSocketPosix::get_socket_address(IP_Address *r_ip, uint16_t *r_port) const {
+	IP_Address temp_ip;
+	uint16_t temp_port;
+
+	IP_Address &ip_ref = r_ip != NULL ? *r_ip : temp_ip;
+	uint16_t &port_ref = r_port != NULL ? *r_port : temp_port;
```
Comment (Faless) [inline drivers/unix/net_socket_posix.cpp:736]: `NULL -> nullptr`

**Review by Faless — CHANGES_REQUESTED** [review]
This is excellent work :1st_place_medal: :trophy: !

Almost good to merge. See requested style changes.

File: core/io/packet_peer_udp.cpp (hunk 6 above)
Comment (dam-coder) [reply core/io/packet_peer_udp.cpp]: done

File: core/io/packet_peer_udp.cpp (hunk 7 above)
Comment (dam-coder) [reply core/io/packet_peer_udp.cpp]: done

File: core/io/stream_peer_tcp.cpp (hunk 8 above)
Comment (dam-coder) [reply core/io/stream_peer_tcp.cpp]: done

File: core/io/packet_peer_udp.cpp (hunk 9 above)
Comment (dam-coder) [reply core/io/packet_peer_udp.cpp]: done

File: core/io/stream_peer_tcp.cpp (hunk 10 above)
Comment (dam-coder) [reply core/io/stream_peer_tcp.cpp]: done

File: core/io/tcp_server.cpp (hunk 11 above)
Comment (dam-coder) [reply core/io/tcp_server.cpp]: done

File: core/io/udp_server.cpp (hunk 12 above)
Comment (dam-coder) [reply core/io/udp_server.cpp]: done

File: drivers/unix/net_socket_posix.cpp (hunk 13 above)
Comment (dam-coder) [reply drivers/unix/net_socket_posix.cpp]: done

**Comment by dam-coder** [comment]
Applied style changes as requested by @Faless .

File: core/io/packet_peer_udp.cpp [hunk 14]
```diff
@@ -328,21 +337,32 @@ int PacketPeerUDP::get_packet_port() const {
...
+	return local_port;
+}
+
 void PacketPeerUDP::set_dest_address(const IP_Address &p_address, int p_port) {
 	ERR_FAIL_COND_MSG(connected, "Destination address cannot be set for connected sockets");
 	peer_addr = p_address;
 	peer_port = p_port;
 }
 
+Error PacketPeerUDP::get_socket_address(IP_Address *r_ip, uint16_t *r_port) const {
```
Comment (Faless) [inline core/io/packet_peer_udp.cpp:352]: this method is unused

File: core/io/packet_peer_udp.h [hunk 15]
```diff
@@ -78,11 +78,13 @@ class PacketPeerUDP : public PacketPeer {
 	Error connect_shared_socket(Ref<NetSocket> p_sock, IP_Address p_ip, uint16_t p_port, UDPServer *ref); // Used by UDPServer
 	void disconnect_shared_socket(); // Used by UDPServer
 	Error store_packet(IP_Address p_ip, uint32_t p_port, uint8_t *p_buf, int p_buf_size); // Used internally and by UDPServer
-	Error connect_to_host(const IP_Address &p_host, int p_port);
+	Error connect_to_host(const IP_Address &p_host, int p_port, int p_local_port = 0);
 	bool is_connected_to_host() const;
+	Error get_socket_address(IP_Address *r_ip, uint16_t *r_port) const;
```
Comment (Faless) [inline core/io/packet_peer_udp.h:83]: this method is unused

File: core/io/tcp_server.cpp [hunk 16]
```diff
@@ -74,6 +76,16 @@ Error TCP_Server::listen(uint16_t p_port, const IP_Address &p_bind_address) {
 	return OK;
 }
 
+Error TCP_Server::get_socket_address(IP_Address *r_ip, uint16_t *r_port) const {
```
Comment (Faless) [inline core/io/tcp_server.cpp:79]: this method is unused

File: core/io/tcp_server.h [hunk 17]
```diff
@@ -49,6 +49,8 @@ class TCP_Server : public Reference {
 
 public:
 	Error listen(uint16_t p_port, const IP_Address &p_bind_address = IP_Address("*"));
+	Error get_socket_address(IP_Address *r_ip, uint16_t *r_port) const;
```
Comment (Faless) [inline core/io/tcp_server.h:52]: this method is unused

File: core/io/udp_server.cpp [hunk 18]
```diff
@@ -112,11 +114,19 @@ Error UDPServer::listen(uint16_t p_port, const IP_Address &p_bind_address) {
 		stop();
 		return err;
 	}
-	bind_address = p_bind_address;
-	bind_port = p_port;
 	return OK;
 }
 
+Error UDPServer::get_socket_address(IP_Address *r_ip, uint16_t *r_port) const {
```
Comment (Faless) [inline core/io/udp_server.cpp:120]: this method is unused

File: core/io/udp_server.h [hunk 19]
```diff
@@ -53,21 +53,19 @@ class UDPServer : public Reference {
...
 
 	Ref<NetSocket> _sock;
-
 	static void _bind_methods();
 
 public:
 	void remove_peer(IP_Address p_ip, int p_port);
 	Error listen(uint16_t p_port, const IP_Address &p_bind_address = IP_Address("*"));
 	Error poll();
+	Error get_socket_address(IP_Address *r_ip, uint16_t *r_port) const;
```
Comment (Faless) [inline core/io/udp_server.h:67]: this method is unused

**Review by Faless — CHANGES_REQUESTED** [review]
see notes on unused methods

File: core/io/packet_peer_udp.cpp (hunk 14 above)
Comment (dam-coder) [reply core/io/packet_peer_udp.cpp]: Removed.

File: core/io/packet_peer_udp.h (hunk 15 above)
Comment (dam-coder) [reply core/io/packet_peer_udp.h]: Removed.

File: core/io/tcp_server.cpp (hunk 16 above)
Comment (dam-coder) [reply core/io/tcp_server.cpp]: Removed.

File: core/io/tcp_server.h (hunk 17 above)
Comment (dam-coder) [reply core/io/tcp_server.h]: Removed.

File: core/io/udp_server.cpp (hunk 18 above)
Comment (dam-coder) [reply core/io/udp_server.cpp]: Removed.

File: core/io/udp_server.h (hunk 19 above)
Comment (dam-coder) [reply core/io/udp_server.h]: Removed.

**Comment by dam-coder** [comment]
Removed unused methods as requested by @Faless.

File: core/io/stream_peer_tcp.cpp [hunk 20]
```diff
@@ -80,6 +82,12 @@ Error StreamPeerTCP::connect_to_host(const IP_Address &p_host, uint16_t p_port)
 
 	_sock->set_blocking_enabled(false);
 
+	if (_sock->bind(IP_Address("*"), p_local_port) != OK) {
```
Comment (Faless) [inline core/io/stream_peer_tcp.cpp:85]: I can foresee this giving us troubles in some dual stack networks.
If we want to be able to bind before connect in TCP (allowing for those rare cases where TCP NAT-traversal works), I think it's safer to expose it as a separate `bind_socket` method that returns an `Error`.

**Review by Faless — CHANGES_REQUESTED** [review]
See notes about TCP/UDP local_port as an argument vs using `udp.listen` and `tcp.bind_socket`.

**Comment by dam-coder** [comment]
I'm sorry but I no longer have the availability to continue working on this PR.
Hope it may be useful... if not, at least I had some fun with it =)

**Comment by Faless** [comment]
I'm sorry to hear that, the review process is quite taxing on everyone, and I think this was good work and definitely useful, it just needed some more thoughts on the workflow. I'll salvage this for `4.0`.

I'm glad you had some fun with it, and hope you would consider contributing again in the future.
